### PR TITLE
SALTO-825: move filter runner to adapter utils

### PR DESCRIPTION
**Pull request by alonstern**
Moved the filterRunner and the relevant interfaces to adapter-utils and used it in all the adapters.

---
_Release Notes_: 
None

### CODE REVIEW

File: packages/adapter-utils/src/filter.ts [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  FilterWith<R, M>[] =>
+      types.filterHasMember<Filter<R>, M>(m, allFilters)
+
+  return {
+    onFetch: async elements => {
+      const filterResults = (await promises.array.series(
+        filtersWith('onFetch').map(filter => () => filter.onFetch(elements))
+      )).filter(isDefined)
+      return concatObjects(
+        filterResults as Record<string, unknown[] | undefined>[]
```
Comment (netama) [inline packages/adapter-utils/src/filter.ts:71]: these casts are not ideal :) I'm wondering if we actually need the ` | void` in all places? what if we just had
```
onFetch(elements: Element[]): Promise<T | void>
```
and kept `T` and `R` everywhere as the non-void variants? would we lose any safeties?

(if we do this then please ignore the earlier comment on the `onFetch` response type)

File: packages/adapter-utils/src/filter.ts [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+      ) as unknown as R
+    },
+    preDeploy: async changes => {
+      await promises.array.series(filtersWith('preDeploy').reverse().map(filter => () => filter.preDeploy(changes)))
+    },
+    onDeploy: async changes => {
+      await promises.array.series(filtersWith('onDeploy').map(filter => () => filter.onDeploy(changes)))
```
Comment (netama) [inline packages/adapter-utils/src/filter.ts:78]: we don't need the `.flat()` part at the end anymore? (I guess this ties to the fact we no longer expect anything in the response, I'm not fully following that change and whether it's related to anything in this PR))

File: packages/adapter-utils/src/filter.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+// Filters run in a specific order and get a mutable list as input which they may modify
+// to affect the overall result as well as the input for subsequent filters.
+// on preDeploy the filters are run in reverse order and are expected to "undo" any relevant change
+// they made in onFetch. because of this, each filter can expect to get in preDeploy a similar
+// value to what it created in onFetch.
+// onDeploy is called in the same order as onFetch and is expected to do basically the same thing
+// that onFetch does but with a different context (on changes instead of on elements)
+// Each filter will be created once and so it may store context between preDeploy and onDeploy.
+// Note that it cannot store context between onFetch and the other callbacks since these run in
+// separate commands
```
Comment (netama) [inline packages/adapter-utils/src/filter.ts:31]: This should probably be updated to reflect `onPostFetch` too (I think it would be easier to document each function on its own, but since this is the existing format if you want to keep it it should probably be mentioned here too). specifically:
1. `onPostFetch` is run after fetch completed for all services, and receives as context all the elements for the env. It should only be used to change references, and should not make any changes that other services might rely on.
2. There is no guarantee on the order in which the `onPostFetch` operations from different adapters are performed, only within each adapter.
3. The filters are run in the same order as `onFetch`.

File: packages/salesforce-adapter/test/filters/filters.test.ts [hunk 4]
```diff
@@ -36,7 +36,8 @@ describe('SalesforceAdapter filters', () => {
       filter = {
         onFetch: mockFunction<(typeof filter)['onFetch']>().mockResolvedValue(),
         preDeploy: mockFunction<(typeof filter)['preDeploy']>().mockResolvedValue(),
-        onDeploy: mockFunction<(typeof filter)['onDeploy']>().mockResolvedValue([]),
+        onDeploy: mockFunction<(typeof filter)['onDeploy']>().mockResolvedValue(),
+        onPostFetch: mockFunction<(typeof filter)['onPostFetch']>().mockResolvedValue(),
```
Comment (netama) [inline packages/salesforce-adapter/test/filters/filters.test.ts:40]: I think we shouldn't propagate `onPostFetch` to adapters that are not using it yet. is there some advantage to adding it in salesforce that I'm missing?

File: packages/adapter-components/test/filter_utils.test.ts [hunk 5]
```diff
@@ -1,103 +0,0 @@
...
-        { configVal: '123 ', promise: p.promise },
-        filters,
-      )
-      const onPostFetchRes = runner.onPostFetch({
-        elementsByAdapter: {},
-        currentAdapterElements: [],
-        progressReporter: { reportProgress: jest.fn() },
-      })
-      await new Promise(resolve => setTimeout(resolve, 2))
-      expect(mockOnPostFetch2).not.toHaveBeenCalled()
```
Comment (netama) [inline packages/adapter-components/test/filter_utils.test.ts:95]: if you're removing this test please add a replacement for it in `adapter-utils`

File: packages/adapter-utils/src/filter.ts [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+export type FilterResult = Record<string, unknown[] | undefined>
+
+export type Filter<T extends FilterResult | void> = Partial<{
+  onFetch(elements: Element[]): Promise<T | void>
```
Comment (netama) [inline packages/adapter-utils/src/filter.ts:35]: shouldn't this just be `Promise<T>`?

File: packages/salesforce-adapter/src/filter.ts [hunk 7]
```diff
@@ -13,33 +13,12 @@
...
-// value to what it created in onFetch.
-// onDeploy is called in the same order as onFetch and is expected to do basically the same thing
-// that onFetch does but with a different context (on changes instead of on elements)
-// Each filter will be created once and so it may store context between preDeploy and onDeploy.
-// Note that it cannot store context between onFetch and the other callbacks since these run in
-// separate commands
-export type Filter = Partial<{
-  onFetch(elements: Element[]): Promise<FilterResult | void>
-  preDeploy(changes: Change[]): Promise<void>
-  onDeploy(changes: Change[]): Promise<(SaveResult | UpsertResult)[]>
```
Comment (netama) [inline packages/salesforce-adapter/src/filter.ts:39]: we don't need the response details anymore?

File: packages/adapter-utils/src/filter.ts (hunk 1 above)
Comment (alonstern) [reply packages/adapter-utils/src/filter.ts:71]: The `| void` in all the other places is so adapters will be able to do `Filter<void>` if they don't need any return value from the filter

File: packages/adapter-utils/src/filter.ts (hunk 2 above)
Comment (alonstern) [reply packages/adapter-utils/src/filter.ts:78]: Yea, because the return values of onDeploy were never used. it is done in this PR because I try here to create a common interface for all the adapters so I needed onDeploy to return void (I could make it return a template but since nobody uses this anyway there was no reason to)

File: packages/salesforce-adapter/test/filters/filters.test.ts (hunk 4 above)
Comment (alonstern) [reply packages/salesforce-adapter/test/filters/filters.test.ts:40]: It wouldn't compile otherwise :S

File: packages/adapter-utils/src/filter.ts (hunk 6 above)
Comment (alonstern) [reply packages/adapter-utils/src/filter.ts:35]: Answered https://github.com/salto-io/salto/pull/2163#discussion_r659326783

File: packages/salesforce-adapter/src/filter.ts (hunk 7 above)
Comment (alonstern) [reply packages/salesforce-adapter/src/filter.ts:39]: Nope, it was not used

File: packages/adapter-utils/src/filter.ts (hunk 3 above)
Comment (alonstern) [reply packages/adapter-utils/src/filter.ts:31]: will be done

File: packages/adapter-components/test/filter_utils.test.ts (hunk 5 above)
Comment (alonstern) [reply packages/adapter-components/test/filter_utils.test.ts:95]: I added tests in `https://github.com/salto-io/salto/pull/2163/files/f0a682acc20a905dfb80fed8f5f643e6929c69cc#diff-049a7166dc9617105aadd1f12486a552a85ebc0eb20034a2edac36318cfd8cc9`
which I think covers the same flows (?)

File: packages/adapter-utils/src/filter.ts (hunk 1 above)
Comment (netama) [reply packages/adapter-utils/src/filter.ts:71]: I understand :) but can't we achieve the same with some default return type and just having all their filters return `void` (which is also allowed)?

File: packages/adapter-utils/src/filter.ts (hunk 2 above)
Comment (netama) [reply packages/adapter-utils/src/filter.ts:78]: got it - was this coordinated with @ori-moisis ?

File: packages/salesforce-adapter/test/filters/filters.test.ts (hunk 4 above)
Comment (netama) [reply packages/salesforce-adapter/test/filters/filters.test.ts:40]: bummer. I'm wondering if we should make it a little harder to add the first `onPostFetch` filter in each adapter though - we should at least keep track of these more carefully. can we make it optional / templated in the interface?

File: packages/adapter-utils/src/filter.ts (hunk 1 above)
Comment (alonstern) [reply packages/adapter-utils/src/filter.ts:71]: I am not sure what the default return type could be if we remove the `| void from the other places

File: packages/adapter-utils/src/filter.ts (hunk 2 above)
Comment (alonstern) [reply packages/adapter-utils/src/filter.ts:78]: Yep

File: packages/salesforce-adapter/test/filters/filters.test.ts (hunk 4 above)
Comment (alonstern) [reply packages/salesforce-adapter/test/filters/filters.test.ts:40]: Is `onPostFetch` different from the rest of the functions in `Filter`? Anyway it is the adapter's responsibility to make sure the functions are called and it's the same with `pre/onDeploy` in the adapter-components' adapters which do not call `pre/onDeploy`

File: packages/adapter-components/test/filter_utils.test.ts (hunk 5 above)
Comment (netama) [reply packages/adapter-components/test/filter_utils.test.ts:95]: IIUCthey don't cover this check (only resolving some promises and making sure the others were not called yet

File: packages/adapter-utils/test/filter.test.ts [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    'onPostFetch',
+  ]).describe('%s', operation => {
+    const operation1 = jest.fn()
+    const operation2 = jest.fn()
+
+    beforeEach(async () => {
+      jest.resetAllMocks()
+
+      const filters = [operation1, operation2]
+        .map(f => () => ({ [operation]: f })) as unknown as FilterCreator<void, {}>[]
```
Comment (netama) [inline packages/adapter-utils/test/filter.test.ts:59]: do we have a way to avoid the `as unknown` here?

File: packages/adapter-utils/src/filter.ts (hunk 6 above)
Comment (netama) [reply packages/adapter-utils/src/filter.ts:35]: my bad, mixed the `| void` in the template arg

File: packages/adapter-utils/src/filter.ts [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+>(
+    opts: T,
+    filterCreators: ReadonlyArray<FilterCreator<R, T>>,
+  ): Required<Filter<R>> => {
+  // Create all filters in advance to allow them to hold context between calls
+  const allFilters = filterCreators.map(f => f(opts))
+
+  const filtersWith = <M extends keyof Filter<R>>(m: M):
+  FilterWith<R, M>[] =>
+      types.filterHasMember<Filter<R>, M>(m, allFilters)
```
Comment (netama) [inline packages/adapter-utils/src/filter.ts:58]: indentation nit: the response type having the same indentation is a little confusing (+ adding any new code below this messes up the following line's indentation). maybe
```suggestion
  const filtersWith = <M extends keyof Filter<R>>(m: M): FilterWith<R, M>[] => (
    types.filterHasMember<Filter<R>, M>(m, allFilters)
  )
```

File: packages/adapter-utils/src/filter.ts [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+  return {
+    onFetch: async elements => {
+      const filterResults = (await promises.array.series(
+        filtersWith('onFetch').map(filter => () => filter.onFetch(elements))
+      )).filter(isDefined)
+      return concatObjects(
+        filterResults as Record<string, unknown[] | undefined>[]
+      ) as unknown as R
```
Comment (netama) [inline packages/adapter-utils/src/filter.ts:67]: I think this cast is actually incorrect - if `R` really is `void` then we shouldn't be calling `concatObjects` in the first place, no?

File: packages/adapter-utils/src/filter.ts [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+    },
```
Comment (netama) [inline packages/adapter-utils/src/filter.ts:68]: another option is to keep the current template args but add another parameter for the aggregate function - something like
```suggestion
  opts: T,
  filterCreators: ReadonlyArray<FilterCreator<R, T>>,
  onFetchAggregator: (results: (R | void)[]) => R | void = () => undefined,
): Required<Filter<R>> => {
  // Create all filters in advance to allow them to hold context between calls
  const allFilters = filterCreators.map(f => f(opts))

  const filtersWith = <M extends keyof Filter<R>>(m: M):
  FilterWith<R, M>[] =>
      types.filterHasMember<Filter<R>, M>(m, allFilters)

  return {
    onFetch: async elements => {
      const filterResults = (await promises.array.series(
        filtersWith('onFetch').map(filter => () => filter.onFetch(elements))
      ))
      return onFetchAggregator(filterResults)
    },
```

File: packages/salesforce-adapter/src/adapter.ts [hunk 12]
```diff
@@ -294,17 +297,18 @@ export default class SalesforceAdapter implements AdapterOperations {
...
         systemFields,
         useOldProfiles: config.useOldProfiles ?? useOldProfiles,
         fetchProfile: this.fetchProfile,
         elementsSource,
       },
-      filterCreators
-    )
+    },
+    filterCreators,
+    concatObjects)
```
Comment (netama) [inline packages/salesforce-adapter/src/adapter.ts:311]: I think we might need the `isDefined` filter you had on the results before we call `concatOjbects`?

File: packages/salesforce-adapter/test/filters/filters.test.ts (hunk 4 above)
Comment (netama) [reply packages/salesforce-adapter/test/filters/filters.test.ts:40]: discussed IRL, should be ok

**Review by netama — APPROVED** [review]
Cool :) I think there might be one more edge case, but looks great

File: packages/salesforce-adapter/src/adapter.ts (hunk 12 above)
Comment (alonstern) [reply packages/salesforce-adapter/src/adapter.ts:311]: IIUC, this is done in https://github.com/salto-io/salto/pull/2163/files/21fa47485cd06a37fa6a172738b9dd27747c8b50#diff-f10e79e63fb50747bdbf9ff75206b95af9eb1a88920da605e981344f3c6eff7dR64
